### PR TITLE
initial hog database backend

### PR DESCRIPTION
**Pull request by ribalba**
This is the first version of the hog backend. 

I still haven't written the migration script for the DB as the schema might change so you will need to create a new DB.

### CODE REVIEW

File: frontend/js/hog-details.js [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+                } catch (err) {
+                    showNotification('Could not get data from API', err);
+                    return;
+                }
+            }, 1000);
+
+        }
+
+        function focusOnBar(dataIndex) {
+            let zoomFactor = 4;
```
Comment (ArneTR) [inline frontend/js/hog-details.js:285]: all const?

File: frontend/js/hog-details.js [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+
+        function handleZoomEvent(){
+            let zoomTimeout;
+            $('#table-loader').addClass('active');
+
+            clearTimeout(zoomTimeout);
+
+            zoomTimeout = setTimeout(async function() {
+                let dataZoomOption = myChart.getOption().dataZoom[0];
```
Comment (ArneTR) [inline frontend/js/hog-details.js:125]: can apart from firstvalue and lastvalue not all be const?

**Review by ArneTR — CHANGES_REQUESTED** [review]

**Comment by ArneTR** [comment]
Very nice and clean code. I found only some remarks of which many are open to discuss.

The main discussion point I put separately: Shall we call the endpoints and internal namings still hog, or shall we here also move to power-hog?

An argument can be made that in the future there can be other hogs, so keeping the endpoint generic might not be the worst idea ...?

File: api/api.py [hunk 3]
```diff
@@ -525,6 +547,340 @@ async def get_jobs():
...
+@app.post('/v1/hog/add')
+async def hog_add(measurements: List[HogMeasurement]):
+
+    for measurement in measurements:
+        decoded_data = base64.b64decode(measurement.data)
+        decompressed_data = zlib.decompress(decoded_data)
+        measurement_data = json.loads(decompressed_data.decode())
+
+        try:
+            _ = Measurement(**measurement_data)
```
[the remark this answers is not in the document]
Comment (ribalba) [reply api/api.py:567]: I couldn't find a way to decompress the data string for FastAPI error handling to work. Or what do you mean with the 422 validation?

Good point now also returns the exception.

File: api/api.py [hunk 4]
```diff
@@ -525,6 +547,340 @@ async def get_jobs():
...
+
+        # We remove the coalitions as we don't want to save all the data in hog_measurements
+        coalitions = measurement_data['coalitions'].copy()
+        del measurement_data['coalitions']
+        del measurement.data
+
+        cpu_energy_data = {}
+        if 'ane_energy' in measurement_data['processor']:
+            cpu_energy_data = {
+                'combined_power': int(measurement_data['processor'].get('combined_power', 0) * measurement_data['elapsed_ns'] / 1_000_000_000.0),
```
[the remark this answers is not in the document]
Comment (ribalba) [reply api/api.py:583]: to get mJ from mW

totally right. Should be combined_energy as the DB field is.

File: api/api.py [hunk 5]
```diff
@@ -525,6 +547,340 @@ async def get_jobs():
...
+                'cpu_energy': int(measurement_data['processor'].get('cpu_energy', 0)),
+                'gpu_energy': int(measurement_data['processor'].get('gpu_energy', 0)),
+                'ane_energy': int(measurement_data['processor'].get('ane_energy', 0)),
+                'energy_impact': measurement_data['all_tasks'].get('energy_impact'),
+            }
+        elif 'package_joules' in measurement_data['processor']:
+            # Intel processors report in joules/ watts and not mJ
+            cpu_energy_data = {
+                'combined_power': int(measurement_data['processor'].get('package_joules', 0) * 1_000),
```
[the remark this answers is not in the document]
Comment (ribalba) [reply api/api.py:592]: we save everything in mJ (millijoule) but intel processors give us joules. I made the comment more explicit

File: api/api.py [hunk 6]
```diff
@@ -525,6 +547,340 @@ async def get_jobs():
...
+                'cpu_energy': int(measurement_data['processor'].get('cpu_joules', 0) * 1_000),
+                'gpu_energy': int(measurement_data['processor'].get('igpu_watts', 0) * measurement_data['elapsed_ns'] / 1_000_000_000.0 * 1_000),
```
[the remark this answers is not in the document]
Comment (ribalba) [reply api/api.py:594]: to get from watts to millijoule.

File: api/api.py [hunk 7]
```diff
@@ -525,6 +547,340 @@ async def get_jobs():
...
+        GROUP BY
+            name
+        ORDER BY
+            total_energy_impact DESC
+        LIMIT 100;
+    """
+
+    data = DB().fetch_all(query, (measurements_id_start, measurements_id_end,))
+
+    if data is None:
```
[the remark this answers is not in the document]
Comment (ribalba) [reply api/api.py:795]: - I refactored the code to make it clearer what it does. 
- I also didn't know that fetch_all always returns a list. We have that check quite a few times:
```
    data = DB().fetch_all(query, (run_id,))
    if data is None or data == []:
        return Response(status_code=204) # No-Content
```
shall I refactor this everywhere?

- I like `if x is None:` more than `if not x:`. Maybe we should define a std for our code base?

File: api/api.py [hunk 8]
```diff
@@ -525,6 +547,340 @@ async def get_jobs():
...
+            AND c.measurement BETWEEN %s AND %s
+        GROUP BY
+            t.name
+        ORDER BY
+            total_energy_impact DESC;
+    """
+
+    tasks_data = DB().fetch_all(query, (coalition_name, measurements_id_start,measurements_id_end,))
+
+    if tasks_data is None:
```
[the remark this answers is not in the document]
Comment (ribalba) [reply api/api.py:850]: What do you mean with `==`?

File: api/api.py [hunk 9]
```diff
@@ -525,6 +547,340 @@ async def get_jobs():
...
+            name
+        ORDER BY
+            total_energy_impact DESC
+        LIMIT 100;
+    """
+
+    coalitions_data = DB().fetch_all(query, (coalition_name, measurements_id_start, measurements_id_end,))[0]
+
+    if coalitions_data is None:
+        coalitions_data = []
```
[the remark this answers is not in the document]
Comment (ribalba) [reply api/api.py:876]: If it is empty it is ok? I want the list to be empty if there is no data

File: api/object_specifications.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
+from pydantic import BaseModel
+from typing import List, Dict, Optional
+
+class Task(BaseModel):
+    name: str
+    cputime_ns: int
+    timer_wakeups: List
+    diskio_bytesread: int = 0
```
[the remark this answers is not in the document]
Comment (ribalba) [reply api/object_specifications.py:8]: See comment below: https://docs.pydantic.dev/latest/migration/#required-optional-and-nullable-fields

If you have 
```
diskio_bytesread: Optional[int]
```
the key is still required. 

In this case I wanted to also have the `diskio_bytesread` set to 0 as I was planing on using the validated object at some stage. Added the optional keyword to make it clearer.

File: docker/structure.sql [hunk 11]
```diff
@@ -206,3 +206,74 @@ CREATE TRIGGER timeline_projects_moddatetime
...
+    created_at timestamp with time zone DEFAULT now(),
+    updated_at timestamp with time zone
+);
+CREATE TRIGGER hog_coalitions_moddatetime
+    BEFORE UPDATE ON hog_coalitions
+    FOR EACH ROW
+    EXECUTE PROCEDURE moddatetime (updated_at);
+
+CREATE INDEX idx_coalition_energy_impact ON hog_coalitions(energy_impact);
+CREATE INDEX idx_coalition_name ON hog_coalitions(name);
```
[the remark this answers is not in the document]
Comment (ribalba) [reply docker/structure.sql:255]: I think this is a difficult one to spot. I know that sometimes a `group by` is done by sorting. I set an index here to speed up the `group by name`. I don't know if the DB is smart enough to switch to a filter if we set a hash index here. Good little test to set up :)

File: frontend/js/hog-details.js [hunk 12]
```diff
@@ -0,0 +1,321 @@
+$(document).ready(function () {
+    var mData
```
[the remark this answers is not in the document]
Comment (ribalba) [reply frontend/js/hog-details.js:2]: totally

File: api/api.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ArneTR) [reply api/api.py]: @ribalba Remeber that data can be None in case where the DB is not available or the Select query is broken

**Review by ArneTR — CHANGES_REQUESTED** [review]

**Pull request merged**